### DISCUSSION TITLE
Map from a Map<Integer, String>? ==> ( No property named "1" exists in source parameter(s) ....)

### DISCUSSION THREAD

**Discussion by jeffkni** [mapstruct/mapstruct#3148] [General]
I have a `Map<Integer, String>` that I would like to map to a 'bean'. 

For example:

Target bean (generic example):
```java
package com.mycompany.app;
public class MyThing {
    private String a;
    private int b;
    public void setA(String a) { this.a = a; }
    public void setB(int b) { this.b = b; }
    public String getA() { return this.a; }
    public int getB() { return this.b; }
}
```

Mapper:
```java
package com.mycompany.app;
import java.util.Map;
import org.mapstruct.Mapper;
import org.mapstruct.Mapping;
import com.mycompany.app.MyThing;

@Mapper
public interface MyMapper {
      // These mappings don't seem legal. What's the legal version? 
     // What I want in codegen is something like: 
     // target.setA(String.valueOf(map.get(1)))
     // target.setB(Integer.valueOf(map.get(2)))
    @Mapping(source = 1, target = "a")
    @Mapping(source = 2, target = "b") 
    MyThing map(Map<Integer, Object> dict);
}
```

Usage example:
```java
package com.mycompany.app;
import com.mycompany.app.MyMapperImpl;
import com.mycompany.app.MyThing;
import java.util.Map;
import java.util.HashMap;

public class App
{
    public static void main( String[] args )
    {
        Map<Integer, Object> dict = new HashMap<Integer, Object>();
        dict.put(1, "here is your string");
        dict.put(2, 5555);

        // Map it 
        MyThing thing = new MyMapperImpl().map(dict);

        System.out.println("Hello A: "  + thing.getA()); // Expected:  Hello A: here is your string
        System.out.println("Hello B " + thing.getB()); // Expected:  Hello : 5555
    }
}
```

It seems like this should be straight forward and I'd expect it to work just like a `Map<String, String>` which works fine but this gives me:

>  No property named "1" exists in source parameter(s). Did you mean "empty"?

**Comment by jknight** [comment]
Looks like Mapping with a Map must be a `Map<String` ? 


```
DiagnosticDescriptor: WARNING Issue001Mapper.java:28 The Map parameter "source" cannot be used for property mapping. 
It must be typed with Map<String ???> but it was typed with Map<Integer,Object>.]
```

@filiphr so many guards against my Map having an Integer as key :) 

```java
  private boolean hasStringMapSignature() {
        if ( isMapType() ) {
            List<Type> typeParameters = getTypeParameters();
            if ( typeParameters.size() == 2 && typeParameters.get( 0 ).isString() ) {
                return true;
            }
        }
        return false;
    }
```

and for example this.It looks like it's really not meant to do what I'm trying to do :) Any other ways to accomplish this?

```java
    static PresenceCheckAccessor mapContainsKey(String propertyName) {
        return suffix( ".containsKey( \"" + propertyName + "\" )" );
    }
```

**Comment by filiphr** [comment]
MapStruct currently only supports mapping from a `Map<String, ?>` to a Bean it doesn't support other types of maps. This is explicitly explained in the [Mapping Map to Bean](https://mapstruct.org/documentation/stable/reference/html/#mapping-map-to-bean) in our documentation. 

The idea behind this was that you can map Objects from a map and objects only have strings as properties. If you want to use some other special maps you'll need to use a custom way of mapping it. 

An alternative would be that we support some kind of syntax that would allow extracting things from a map using the `source` attribute. e.g. `source = "map[1]`. I think that we have a request for collections for this, but I none of the MapStruct team members is working on this.

**Reply by katerynasvitlychna** [reply]
Hi @filiphr . I'm facing the same problem. How come it wasn't an issue in earlier versions? I checked on 1.4.2.Final and it worked fine. But after an update to 1.6.0 it gives warning `It must be typed with Map<String, ???> but it was typed with Map<Long,String>.`

**Reply by filiphr** [reply]
@katerynasvitlychna, I'll need some more information about your problem. The support for mapping `Map<String, ???>` to bean is there since 1.5 (https://mapstruct.org/documentation/dev/reference/html/#mapping-map-to-bean).

**Reply by katerynasvitlychna** [reply]
Yes, it works for Map<String, Object>. But in my case, I have Map<Long, String>. It used to work on older versions, so I'm a little bit confused as to why it doesn't anymore

**Reply by filiphr** [reply]
@katerynasvitlychna I think that you are having some other problem. Can you please create a different discussion and provide what you think used to work and what isn't working now. 

Mapping `Map<String, ???>` to bean did not work in 1.4 for sure.

**Reply by katerynasvitlychna** [reply]
@filiphr of course
